internal/hostport/hostport_nftables.go

```diff
@@ -0,0 +1,308 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hostport
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base32"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+
+	"sigs.k8s.io/knftables"
+)
+
+const (
+	// our nftables table.
+	hostPortsTable string = "crio-hostports"
+
+	// maps and sets referred to from HostportManager.
+	hostPortsMap   string = "hostports"
+	hostIPPortsMap string = "hostipports"
+	hairpinSet     string = "hairpins"
+)
+
+type hostportManagerNFTables struct {
+	nft    knftables.Interface
+	family knftables.Family
+	mu     sync.Mutex
+}
+
+// newHostportManagerNFTables creates a new nftables HostPortManager.
+func newHostportManagerNFTables(family knftables.Family) (*hostportManagerNFTables, error) {
+	nft, err := knftables.New(family, hostPortsTable)
+	if err != nil {
+		return nil, err
+	}
+
+	return &hostportManagerNFTables{
+		nft:    nft,
+		family: family,
+	}, nil
+}
+
+func (hm *hostportManagerNFTables) Add(id, name, podIP string, hostportMappings []*PortMapping) (err error) {
+	// Ensure atomicity for nftables operations
+	hm.mu.Lock()
+	defer hm.mu.Unlock()
+
+	tx := hm.nft.NewTransaction()
+	ensureHostPortsTable(tx, hm.family)
+
+	// Add map/set elements to tx for all mappings in hostportMappings. We add a
+	// comment to each element based on the sandbox ID, so we can match them up to
+	// this pod in Remove().
+	comment := hashSandboxID(id)
+
+	for _, pm := range hostportMappings {
+		protocol := strings.ToLower(string(pm.Protocol))
+		hostPort := strconv.Itoa(int(pm.HostPort))
+		containerPort := strconv.Itoa(int(pm.ContainerPort))
+
+		if pm.HostIP == "" || pm.HostIP == "0.0.0.0" || pm.HostIP == "::" {
+			tx.Add(&knftables.Element{
+				Map: hostPortsMap,
+				Key: []string{
+					protocol, hostPort,
+				},
+				Value: []string{
+					podIP, containerPort,
+				},
+				Comment: &comment,
+			})
+		} else {
+			tx.Add(&knftables.Element{
+				Map: hostIPPortsMap,
+				Key: []string{
+					pm.HostIP, protocol, hostPort,
+				},
+				Value: []string{
+					podIP, containerPort,
+				},
+				Comment: &comment,
+			})
+		}
+	}
+
+	// If we added any hostport mappings, add a hairpinning mapping.
+	if tx.NumOperations() != 0 {
```

Comment:
doesn't this also take into account the operations in `ensureHostPortsTable(tx, hm.family)` ?

is not that you always need it if  len(hostportMappings) > 0?

Reply:
hm... you're right, this is wrong.

I think in an earlier stage of the refactoring, `hostportMappings` still had both IP families in it, so `len(hostportMappings) > 0` wouldn't work, but you're right, at this point that's what I need to be checking.

Reply:
(actually, we don't even need to check, because the MetaHostportManager wouldn't have called us if there were no mappings)